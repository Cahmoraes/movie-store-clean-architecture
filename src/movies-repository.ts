import { ChildrenMovie } from './movie/children-movie'
import { Movie } from './movie/movie'
import { NewMovie } from './movie/new-movie'
import { RegularMovie } from './movie/regular-movie'
import { MoviesRepository } from './repository/movie-repository'

interface MovieRootDto {
  [movieId: string]: MovieDto
}

interface MovieDto {
  title: string
  code: string
}

export enum MovieType {
  REGULAR = 'regular',
  NEW = 'new',
  CHILDREN = 'children',
}

export class MoviesRepositoryImpl implements MoviesRepository {
  constructor(private moviesDto: MovieRootDto) {}

  movieForId(aMovieId: string): Movie {
    const movieDto = this.moviesDto[aMovieId]
    return this.create(movieDto)
  }

  private create(movieDto: MovieDto): Movie {
    switch (movieDto.code) {
      case MovieType.REGULAR:
        return new RegularMovie(movieDto.code, movieDto.title)
      case MovieType.NEW:
        return new NewMovie(movieDto.code, movieDto.title)
      case MovieType.CHILDREN:
        return new ChildrenMovie(movieDto.code, movieDto.title)
      default:
        throw new Error('Invalid movie code')
    }
  }
}
