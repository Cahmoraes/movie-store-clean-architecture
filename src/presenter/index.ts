export interface Presenter<Type> {
  print(statement: Type): string
}
