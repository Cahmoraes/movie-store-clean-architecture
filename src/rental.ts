export class Rental {
  constructor(readonly movieID: string, readonly days: number) {}
}
