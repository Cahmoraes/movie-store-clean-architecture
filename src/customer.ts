import { Rental } from './rental'

interface CustomerArgDto {
  name: string
  rentals: {
    movieID: string
    days: number
  }[]
}

export class Customer {
  private _name: string
  private _rentals: Rental[]

  constructor(data: CustomerArgDto) {
    this._name = data.name
    this._rentals = data.rentals.map(
      (rental) => new Rental(rental.movieID, rental.days),
    )
  }

  get name() {
    return this._name
  }

  get rentals() {
    return this._rentals
  }
}
