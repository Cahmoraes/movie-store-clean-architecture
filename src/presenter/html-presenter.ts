import { Presenter } from '.'
import { StatementData } from '../statement'

export class HTMLPresenter implements Presenter<StatementData> {
  print(statementData: StatementData): string {
    let result = `<h1>Rental Record for <em>${statementData.customerName}</em></h1>\n`
    result += '<table>\n'
    for (const rental of statementData.rentals) {
      result += `  <tr><td>${
        rental.movie.title
      }</td><td>${rental.movie.calculateAmountByRental(rental)}</td></tr>\n`
    }
    result += '</table>\n'
    result += `<p>Amount owed is <em>${statementData.amount}</em></p>\n`
    result += `<p>You earned <em>${statementData.frequentRenterPoints}</em> frequent renter points</p>\n`
    return result
  }
}
