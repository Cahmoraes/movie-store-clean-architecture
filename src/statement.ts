import { Customer } from './customer'
import { Movie } from './movie/movie'
import { MoviesRepositoryImpl } from './movies-repository'
import { Rental } from './rental'
import { Presenter } from './presenter/index'
import { MoviesRepository } from './repository/movie-repository'

interface CustomerDto {
  name: string
  rentals: { movieID: string; days: number }[]
}

type TransformedRental = Rental & { movie: Movie }

export interface StatementData {
  customerName: string
  rentals: TransformedRental[]
  amount: number
  frequentRenterPoints: number
}

export class Statement {
  private customer: Customer
  private moviesRepository: MoviesRepository
  private presenter: Presenter<StatementData>

  constructor(
    customer: CustomerDto,
    movies: MoviesRepository,
    presenter: Presenter<StatementData>,
  ) {
    this.customer = new Customer(customer)
    this.moviesRepository = movies
    this.presenter = presenter
  }

  public calculate() {
    const statementData = this.createStatementData()
    return this.presenter.print(statementData)
  }

  private createStatementData() {
    return {
      customerName: this.customer.name,
      rentals: this.customer.rentals.map((rental) => ({
        ...rental,
        movie: this.moviesRepository.movieForId(rental.movieID),
      })),
      amount: this.amount,
      frequentRenterPoints: this.frequentRenterPoints,
    }
  }

  private get frequentRenterPoints() {
    let frequentRenterPoints = 0
    for (const rental of this.customer.rentals) {
      frequentRenterPoints++
      if (this.movieForId(rental.movieID).hasBonusForTwoDaysRelease(rental))
        frequentRenterPoints++
    }
    return frequentRenterPoints
  }

  private movieForId(aMovieId: string) {
    return this.moviesRepository.movieForId(aMovieId)
  }

  private get amount() {
    return this.customer.rentals.reduce(
      (total, rental) =>
        total + this.movieForId(rental.movieID).calculateAmountByRental(rental),
      0,
    )
  }
}
