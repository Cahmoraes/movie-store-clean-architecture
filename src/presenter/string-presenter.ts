import { Presenter } from '.'
import { StatementData } from '../statement'

export class StringPresenter implements Presenter<StatementData> {
  print(statementData: StatementData): string {
    let result = `Rental Record for ${statementData.customerName}\n`
    for (let rental of statementData.rentals) {
      result += `\t${
        rental.movie.title
      }\t${rental.movie.calculateAmountByRental(rental)}\n`
    }
    result += `Amount owed is ${statementData.amount}\n`
    result += `You earned ${statementData.frequentRenterPoints} frequent renter points\n`
    return result
  }
}
