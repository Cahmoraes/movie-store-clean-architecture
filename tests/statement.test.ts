import { test, expect, beforeEach } from 'vitest'
import { Statement } from '../src/statement'
import customerDto from '../repository/customers.json'
import moviesDto from '../repository/movies.json'
import { StringPresenter } from '../src/presenter/string-presenter'
import { MoviesRepositoryImpl } from '../src/movies-repository'
import { HTMLPresenter } from '../src/presenter/html-presenter'

let moviesRepository: MoviesRepositoryImpl

beforeEach(() => {
  moviesRepository = new MoviesRepositoryImpl(moviesDto)
})

test('Deve exibir relatório textual', () => {
  const result = new Statement(
    customerDto,
    moviesRepository,
    new StringPresenter(),
  ).calculate()

  expect(result).toMatchSnapshot()
  console.log(result)
})

test('Deve exibir relatório HTML', () => {
  const result = new Statement(
    customerDto,
    moviesRepository,
    new HTMLPresenter(),
  ).calculate()

  expect(result).toMatchSnapshot()
  console.log(result)
})
