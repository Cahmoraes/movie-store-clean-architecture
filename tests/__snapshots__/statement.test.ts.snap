// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Deve exibir relatório HTML 1`] = `
"<h1>Rental Record for <em>martin</em></h1>
<table>
  <tr><td>Ran</td><td>3.5</td></tr>
  <tr><td>Trois Couleurs: Bleu</td><td>15</td></tr>
</table>
<p>Amount owed is <em>18.5</em></p>
<p>You earned <em>3</em> frequent renter points</p>
"
`;

exports[`Deve exibir relatório textual 1`] = `
"Rental Record for martin
	Ran	3.5
	Trois Couleurs: Bleu	15
Amount owed is 18.5
You earned 3 frequent renter points
"
`;
